src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -396,6 +376,7 @@ ArrayDeque<ParallelBlockStart> leastCommonAncestor(@NonNull final Set<FlowNode>
 
                 // Welp we hit the end of a branch
                 if (!blockStartIterator.hasNext()) {
+                    // This case is why we use ExpandedEnclosingBlockIterator instead of FlowNode.iterateEnclosingBlocks - otherwise branch ordering is not preserved.
```

Comment:
This is the second issue. Otherwise, heads for completed branches get skipped here to start with and are then examined separately later, so the completed branches and incomplete branches were kind of in two separate groups that got concatenated.